pkg/private/pkg_files.bzl

```diff
@@ -0,0 +1,321 @@
+# Copyright 2021 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Internal functions for processing pkg_file* instances.
+
+Concepts and terms:
+
+  DestFile: A provider holding the source path, attributes and other
+            information about a file that should appear in the package.
+            When attributes are empty in DestFile, we let the package
+            tool decide their values.
+
+  content map: The map of destination paths to DestFile instances. Note that
+               several distinct destinations make share the same source path.
+               Attempting to insert a duplicate entry in the content map is
+               an error, because it means you are collapsing files together.
+               We may want to relax this in the future if their DestFiles
+               are equal.
+
+  manifest: The file which represents the content map. This is generated
+            by rule implementations and passed to the build_*.py helpers.
+"""
+
+load("//:path.bzl", "compute_data_path", "dest_path")
+load(
+    "//:providers.bzl",
+    "PackageArtifactInfo",
+    "PackageDirsInfo",
+    "PackageFilegroupInfo",
+    "PackageFilesInfo",
+    "PackageSymlinkInfo",
+)
+
+# Possible values for entry_type
+ENTRY_IS_FILE = 0  # Entry is a file: take content from <src>
+ENTRY_IS_LINK = 1  # Entry is a symlink: dest -> <src>
+ENTRY_IS_DIR = 2  # Entry is an emptry dir
+ENTRY_IS_TREE = 3  # Entry is a tree artifact: take tree from <src>
+
+_DestFile = provider(
+    doc = """Information about each destination in the final package.""",
+    fields = {
+        "src": "source file",
+        "mode": "mode, or empty",
+        "user": "user, or empty",
+        "group": "group, or empty",
+        "link_to": "path to link to. src must not be set",
+        "entry_type": "int. See ENTRY_IS_* values above.",
+        "origin": "target which added this",
+    },
+)
```

Comment:
WRT https://github.com/bazelbuild/rules_pkg/pull/367#discussion_r661779617

Consider documenting type requirements here, specifically integers vs strings.

Reply:
Ack. I can expand that when I add the capability for both string and int user/groups.